pontoon/translate/static/js/translate/TranslationEditor.es6

```diff
@@ -0,0 +1,356 @@
+import {notify} from './utils.js';
+
+
+/**
+ * Root Component for the entire translation editor.
+ */
+export default class TranslationEditor extends React.Component {
+  constructor(props) {
+    super(props)
+    this.state = {
+      loaded: false,
+      sidebarOpen: true,  // Temporarily defaults to open for dev.
+      entities: [],
+      selectedIndex: 0,
+    };
+  }
+
+  render() {
+    return (
+      <div id="translation-editor">
+        <aside id="sidebar"
+               className={classNames('advanced', {open: this.state.sidebarOpen})}
+               ref="sidebar">
+          <EntityList entities={this.props.entities}
+                      project={this.props.project}
+                      selectedIndex={this.state.selectedIndex}
+                      onSelectEntity={this.selectEntity.bind(this)} />
+          <EntityDetails entity={this.selectedEntity}
+                         onSelectNextEntity={this.selectNextEntity.bind(this)}
+                         onSelectPreviousEntity={this.selectPreviousEntity.bind(this)}
+                         onSave={this.saveTranslation.bind(this)} />
+        </aside>
+      </div>
+    );
+  }
+
+  get selectedEntity() {
+    return this.props.entities[this.state.selectedIndex];
+  }
+
+  selectEntity(entityKey) {
+    let index = this.props.entities.findIndex((e) => e.pk === entityKey);
+    if (index !== -1) {
+      this.setState({
+        selectedIndex: index,
+      });
+    }
+  }
+
+  saveTranslation(string) {
+    this.props.onSaveTranslation(this.selectedEntity, string);
+  }
+
+  selectNextEntity(e) {
+    e.preventDefault();
+    if (this.state.selectedIndex < this.props.entities.length - 1) {
+      this.setState({
+        selectedIndex: this.state.selectedIndex + 1,
```

Comment:
This should be the `this.setState((props, state) => { return {selectedIndex: state.selectedIndex + 1}; })` form. Otherwise you can run into race conditions.